src/contracts/deposit/integrity-checker/integrity-checker.service.ts

```diff
@@ -0,0 +1,142 @@
+import { Inject, Injectable, LoggerService } from '@nestjs/common';
+import { RepositoryService } from 'contracts/repository';
+import { WINSTON_MODULE_NEST_PROVIDER } from 'nest-winston';
+import { BlockTag } from 'provider';
+import { DepositTree } from '../deposit-tree';
+import {
+  VerifiedDepositEvent,
+  VerifiedDepositEventsCache,
+} from '../interfaces';
+import { DEPOSIT_TREE_STEP_SYNC } from './constants';
+
+@Injectable()
+export class DepositIntegrityCheckerService {
+  finalizedTree = new DepositTree();
+  constructor(
+    @Inject(WINSTON_MODULE_NEST_PROVIDER) private logger: LoggerService,
+    private repositoryService: RepositoryService,
+  ) {}
+
+  /**
+   * Initializes the deposit tree with an initial cache of verified deposit events.
+   * @param {VerifiedDepositEventsCache} initialEventsCache - Cache of verified deposit events to initialize the tree.
+   */
+  public async initialize(initialEventsCache: VerifiedDepositEventsCache) {
+    await this.putEventsToTree(this.finalizedTree, initialEventsCache.data);
+  }
+
+  /**
+   * Inserts a list of finalized verified deposit events into the deposit tree and returns the updated tree.
+   * @param {VerifiedDepositEvent[]} eventsCache - Array of verified deposit events to be added to the tree.
+   * @returns {Promise<DepositTree>} The updated deposit tree after adding the events.
+   */
+  public async putFinalizedEvents(
+    eventsCache: VerifiedDepositEvent[],
+  ): Promise<DepositTree> {
+    await this.putEventsToTree(this.finalizedTree, eventsCache);
+    return this.finalizedTree;
+  }
+
+  /**
+   * Inserts a list of latest verified deposit events into a clone of the deposit tree and returns the cloned tree.
+   * @param {VerifiedDepositEvent[]} eventsCache - Array of verified deposit events to be added to the cloned tree.
+   * @returns {Promise<DepositTree>} The cloned and updated deposit tree after adding the events.
+   */
+  public async putLatestEvents(
```

Comment:
didnt get idea of having to different functions putFinalizedEvents and putLatestEvents